cloudinit/cmd/main.py

```diff
@@ -1122,8 +1144,69 @@ def main(sysv_args=None):
 
             status_parser(parser_status)
             parser_status.set_defaults(action=("status", handle_status_args))
+    else:
+        parser.error("a subcommand is required")
 
     args = parser.parse_args(args=sysv_args)
+    setattr(args, "skip_log_setup", False)
+    if not args.all_stages:
+        return sub_main(args)
+    LOG.info("Running cloud-init in single process mode.")
+
+    # this _must_ be called before sd_notify is called otherwise netcat may
+    # attempt to send "start" before a socket exists
+    sync = socket.SocketSync("local", "network", "config", "final")
+
+    # notify systemd that this stage has completed
+    socket.sd_notify("READY=1")
+    # wait for cloud-init-local.service to start
+    with sync("local"):
```

Comment:
> Are there ordering requirements inherent to cloud-init's different stages?

Yes, these ordering requirements are inherent requirements. This order is already encoded in the init system. Currently `cloud-init-local.service` is ordered before `cloud-init.service` which is before `cloud-config.service` which is before `cloud-final.service`.

> local must be run before network, for example, this would be an ideal place to encode these requirements.

@setharnold Maybe you had something specific in mind that I'm missing?

Reply:
If these have to happen in lockstep order, maybe a variable to show what stage it's on:

```
stage=0
with sync("local"):
    if (stage++ < 0):
        error_and_exit("unexpected stage error")
    # stuff
with sync("network"):
    if (stage++ < 1):
        error_and_exit("network must follow local")
    # stuff
with sync("config"):
    if (stage++ < 2):
        error_and_exit("config must follow network")
    # stuff
with sync # over and over again
...
```

If the ordering is more vague, like "config" requires "local" and doesn't care about "network", then `local_done` and `network_done` and `config_done` and so on boolean variables to mark these completed in turn, then each one could check the status of the phases that it depends upon.

Reply:
> If these have to happen in lockstep order, maybe a variable to show what stage it's on:
> 
> ```
> stage=0
> with sync("local"):
>     if (stage++ < 0):
>         error_and_exit("unexpected stage error")
>     # stuff
> with sync("network"):
>     if (stage++ < 1):
>         error_and_exit("network must follow local")
>     # stuff
> with sync("config"):
>     if (stage++ < 2):
>         error_and_exit("config must follow network")
>     # stuff
> with sync # over and over again
> ...
> ```
> 
> If the ordering is more vague, like "config" requires "local" and doesn't care about "network", then `local_done` and `network_done` and `config_done` and so on boolean variables to mark these completed in turn, then each one could check the status of the phases that it depends upon.

We could add code like this, but it wouldn't be possible for `error_and_exit()` to ever be called, regardless of the order that the messages are received on the unix sockets. This is single threaded code and it only has a single entry point. So in order for the network stage to run, the local stage has to have run.

Reply:
I think I've just got less faith than you in the systemd-enforced ordering, that's all. If you're sure I'm over-thinking it, feel free to leave it out.

Reply:
> I think I've just got less faith than you in the systemd-enforced ordering, that's all. 

Systemd is not the only thing enforcing ordering - it is already enforced in the Python code as well. If systemd triggers stages out of order, earlier stages will not get skipped - if "network" is triggered before "local", this Python code will wait until systemd triggers "local".

The reason is that the ordering of these stages isn't enforced in the definition of the context manager. It is enforced by the order that the context manager is invoked in - since `with sync("local")` is called prior to `with sync("network")` is called, the network stage cannot start until the local stage has completed. No parallalism / async / threading / coroutines / etc are in use; the code simply runs one stage after the next but not until systemd has signaled that it is time for the stage to run.

> If you're sure I'm over-thinking it, feel free to leave it out.

I'm pretty sure that's what is happening, but I want to make sure you're comfortable with / understanding with how it works too. Hopefully the following annotated code helps.

```python
with sync("local"):
   # "start" must have been received on the local socket to enter the context manager
   # (triggered by the one-liner in cloud-init-local.service)

# once the code gets to this point, the "local" stage has completed (and a response was
# sent in the context manager to the shim process in cloud-init-local.service)
with sync("network"):
   # "start" must have been received on the network socket to enter the context manager
   # (triggered by the one-liner in cloud-init-network.service) - in addition to
   # all of the code above being completed

# once the code gets to this point, the "network" stage has completed (and a response 
# was sent in the context manager to the shim process in cloud-init-network.service)
with sync("config"):
   # "start" must have been received on the config socket to enter the context manager
   # (triggered by the one-liner in cloud-config.service) - in addition to
   # the code above being completed

# once the code gets to this point, the "config" stage has completed (and a response 
# was sent in the context manager to the shim process in cloud-config.service)
with sync("final"):
   # "start" must have been received on the final socket to enter the context manager
   # (triggered by the one-liner in cloud-config.service) - in addition to
   # the code above being completed
```
So even if systemd triggers events out of order, cloud-init won't run them out of order. It will just wait until the next stage in the order has completed. 

Please let me know if you have any questions.

Reply:
Oh! now I can see why you're confident :) thanks!